hal/src/boron/network/sara_ncp_client.cpp

```diff
@@ -155,46 +155,68 @@ int SaraNcpClient::initParser(Stream* stream) {
             .commandTerminator(AtCommandTerminator::CRLF);
     parser_.destroy();
     CHECK(parser_.init(std::move(parserConf)));
+    // +CREG: <stat>[,<lac>,<ci>[,<AcTStatus>]]
     CHECK(parser_.addUrcHandler("+CREG", [](AtResponseReader* reader, const char* prefix, void* data) -> int {
         const auto self = (SaraNcpClient*)data;
-        int val[2];
-        int r = CHECK_PARSER_URC(reader->scanf("+CREG: %d,%d", &val[0], &val[1]));
+        int val[4] = {-1,-1,-1,-1};
+        // Parse as direct response (ignoring mode)
+        int r = CHECK_PARSER_URC(reader->scanf("+CREG: %*d,%d,\"%x\",\"%x\",%d", &val[0], &val[1], &val[2], &val[3]));
+        // Reparse as URC
+        if (r <= 1) { r = CHECK_PARSER_URC(reader->scanf("+CREG: %d,\"%x\",\"%x\",%d", &val[0], &val[1], &val[2], &val[3])); }
         CHECK_TRUE(r >= 1, SYSTEM_ERROR_UNKNOWN);
         // Home network or roaming
-        if (val[r - 1] == 1 || val[r - 1] == 5) {
+        if (val[0] == 1 || val[0] == 5) {
             self->creg_ = RegistrationState::Registered;
         } else {
             self->creg_ = RegistrationState::NotRegistered;
         }
         self->checkRegistrationState();
+        // Cellular Global Identity (partial)
+        self->cgi_.location_area_code = static_cast<uint16_t>(val[2]);
+        self->cgi_.cell_id = val[3];
         return 0;
     }, this));
+    // n={0,1} +CGREG: <stat>
+    // n=2     +CGREG: <stat>[,<lac>,<ci>[,<AcT>,<rac>]]
     CHECK(parser_.addUrcHandler("+CGREG", [](AtResponseReader* reader, const char* prefix, void* data) -> int {
         const auto self = (SaraNcpClient*)data;
-        int val[2];
-        int r = CHECK_PARSER_URC(reader->scanf("+CGREG: %d,%d", &val[0], &val[1]));
+        int val[4];
+        // Parse as direct response (ignoring mode)
+        int r = CHECK_PARSER_URC(reader->scanf("+CGREG: %*d,%d,\"%x\",\"%x\",%d", &val[0], &val[1], &val[2], &val[3]));
+        // Reparse as URC
+        if (r <= 1) { r = CHECK_PARSER_URC(reader->scanf("+CGREG: %d,\"%x\",\"%x\",%d", &val[0], &val[1], &val[2], &val[3])); }
```

Comment:
Formatting. Let's not put code within the codeblock on the same line.

This `if` will invalidate the correct result of the previous scanf for `+CREG: <n>,<stat>` responses, causing `stat` to be parsed as `<n>` instead in the second `scanf`.

Reply:
I forgot to auto format code here, so it will be broken out into multiple lines as soon as I do.

Since, I'm ignoring the first integer, I think the URC will not actually parse any values, so I believe the correct thing to do is check `if (0 == r)` and that should alleviate your concern.

Reply:
That should probably work, yes. It's not a concern though, this is a breaking change.

Reply:
What is "this"? The current behavior of the UrcParser, or checking for `r==0`?

Reply:
Currently checking for `r <= 1` will introduce a bug where `+CREG: <n>,<stat>` responses would get reparsed as URC, which has a different format (`+CREG: <stat>,...`) and [later](https://github.com/particle-iot/device-os/pull/1759/files/a2cf9b9cdb8a892dbdac2ee0ea5b33573f793a15#diff-b3f74fd1e22ae53aae264584d668ea1cR190) we will be erroneously checking `<n>` instead of `<stat>`.

Reply:
This is all fixed by testing against `0 == r`, as we discussed, correct?

Reply:
> That should probably work, yes.

Reply:
please confirm the new logic